daemon/endpoint.go

```diff
@@ -160,24 +159,26 @@ func (d *Daemon) createEndpoint(epTemplate *models.EndpointChangeRequest, id str
 		return PutEndpointIDFailedCode, err
 	}
 
-	// If the endpoint has no labels, give the endpoint a special identity with
-	// label reserved:init so we can generate a custom policy for it until we
-	// get its actual identity.
-	if len(addLabels) == 0 {
+	addLabels := labels.NewLabelsFromModel(lbls)
+
+	if len(addLabels) > 0 {
+		addLabels, _, ok := checkLabels(addLabels, nil)
+		if !ok {
+			return PutEndpointIDInvalidCode, fmt.Errorf("No valid label")
+		}
+		if lbls := addLabels.FindReserved(); lbls != nil {
+			return PutEndpointIDInvalidCode, fmt.Errorf("Not allowed to add reserved labels: %s", lbls)
+		}
+	} else {
+		// If the endpoint has no labels, give the endpoint a special identity with
+		// label reserved:init so we can generate a custom policy for it until we
+		// get its actual identity.
 		addLabels = labels.Labels{
 			labels.IDNameInit: labels.NewLabel(labels.IDNameInit, "", labels.LabelSourceReserved),
 		}
 	}
 
-	code, err := d.updateEndpointLabels(id, addLabels, labels.Labels{})
-	if err != nil {
-		// XXX: Why should the endpoint remain in this case?
-		log.WithFields(logrus.Fields{
-			logfields.EndpointID:     id,
-			logfields.IdentityLabels: logfields.Repr(addLabels),
-		}).WithError(err).Error("Could not add labels while creating an ep")
-		return code, err
-	}
+	ep.SetIdentityLabels(d, addLabels)
```

Comment:
Move this up to before `endpointmanager.AddEndpoint()` so we never even expose the endpoint before it has the init identity.

Reply:
Done.